modyn/supervisor/internal/triggers/ensembletrigger.py

```diff
@@ -0,0 +1,230 @@
+        need the data to maintain their state.
+        Therefore we have to ensure right after a trigger, that all next trigger indexes in the cache are
+        in the future wr.t. our current trigger index. We can simply do that by fetching the next trigger indexes
+        from the subtriggers until all subtriggers have a next trigger index in the future.
+    """
+
+    def __init__(self, config: EnsembleTriggerConfig):
+        self.config = config
+        self.subtriggers = self._create_subtriggers(config)
+
+        self.last_inform_decisions: dict[str, bool] = {}
+        """Leftover decisions from the last inform call of the subtriggers that
+        didn't yield a trigger."""
+
+    def init_trigger(self, context: TriggerContext) -> None:
+        for trigger in self.subtriggers.values():
+            trigger.init_trigger(context)
+
+    def inform(
+        self,
+        new_data: list[tuple[int, int, int]],
+        log: TriggerPolicyEvaluationLog | None = None,
+    ) -> Generator[int, None, None]:
+        subtrigger_generators = {
+            trigger_name: trigger.inform(new_data, log) for trigger_name, trigger in self.subtriggers.items()
+        }
+
+        # Indicates whether the ensemble trigger has passed a triggering index
+        # with processing_head for a subtrigger since the last trigger.
+        # Will be reset after every trigger.
+        subtrigger_decision_cache = dict(
+            {trigger_name: False for trigger_name in self.subtriggers},
+            **self.last_inform_decisions,
+        )
+
+        # Cache for the next trigger index of each subtrigger.
+        # None indicates that the subtrigger has no more triggers because the generator is exhausted.
+        # -1 indicates that the last triggering index was in the last batch or there was no trigger yet.
+        next_subtrigger_index_cache: dict[str, int | None] = {trigger_name: -1 for trigger_name in self.subtriggers}
+
+        # The current candidate index for the next trigger.
+        # If the aggregation function returns True, we trigger. If not,
+        # the processing head is still increased to the next triggering
+        # index of a subtrigger. Indexes are based on the new_data batch.
+        processing_head = -1
+
+        self._update_outdated_next_trigger_indexes(
+            processing_head,
+            subtrigger_generators,
+            next_subtrigger_index_cache,
+        )
+
+        while True:
+            next_trigger_idx = self._find_next_trigger_index(
+                processing_head,
+                new_data,
+                subtrigger_decision_cache,
+                next_subtrigger_index_cache,
+                log,
+            )
+            if next_trigger_idx is None:
+                # complete remaining inform calls regardless of the decision, any of the decision will lead to a
+                # non triggering aggregation result
```

Comment:
I don't understand this - how do we complete something here, if we immediately break out of the loop?

Reply:
Actually, I forgot to implement what I was describing with this comment. We always have to let all subtrigger see all data as they need it for their internal state. After a non trigger and after a trigger we have to consume all generator indexes up to the newest processing head.